client/app/pages/queries/QuerySource.jsx

```diff
@@ -0,0 +1,222 @@
+    };
+  }, []);
+
+  const reloadSchema = useCallback(
+    (refresh = undefined) => {
+      const refreshToken = Math.random()
+        .toString(36)
+        .substr(2);
+      refreshSchemaTokenRef.current = refreshToken;
+      getSchema(dataSource, refresh).then(data => {
+        if (refreshSchemaTokenRef.current === refreshToken) {
+          setSchema(data);
+        }
+      });
+    },
+    [dataSource]
+  );
+
+  useEffect(() => {
+    reloadSchema();
+  }, [reloadSchema]);
+
+  useEffect(() => {
+    recordEvent("view_source", "query", query.id);
+  }, [query.id]);
+
+  const handleDataSourceChange = useCallback(
+    dataSourceId => {
+      if (dataSourceId) {
+        try {
+          localStorage.setItem("lastSelectedDataSourceId", dataSourceId);
+        } catch (e) {
+          // `localStorage.setItem` may throw exception if there are no enough space - in this case it could be ignored
+        }
+      }
+      if (query.data_source_id !== dataSourceId) {
+        recordEvent("update_data_source", "query", query.id, { dataSourceId });
+        const newQuery = clone(query);
+        newQuery.data_source_id = dataSourceId;
+        newQuery.latest_query_data = null;
+        newQuery.latest_query_data_id = null;
+        setQuery(newQuery);
+        // TODO: Save if not new
+      }
+    },
+    [query]
+  );
+
+  useEffect(() => {
+    // choose data source id for new queries
+    if (dataSourcesLoaded && query.isNew()) {
+      const firstDataSourceId = dataSources.length > 0 ? dataSources[0].id : null;
+      handleDataSourceChange(
+        chooseDataSourceId(
+          [query.data_source_id, localStorage.getItem("lastSelectedDataSourceId"), firstDataSourceId],
+          dataSources
+        )
+      );
+    }
+  }, [query, dataSourcesLoaded, dataSources, handleDataSourceChange]);
+
+  return (
+    <div className="query-page-wrapper">
+      <div className="container">
+        <QueryPageHeader query={query} sourceMode onChange={setQuery} />
+      </div>
+      <main className="query-fullscreen">
+        <nav>
+          <div className="editor__left__data-source">
+            <Select
+              className="w-100"
+              placeholder="Choose data source..."
+              value={dataSource ? dataSource.id : undefined}
+              disabled={!query.can_edit || !dataSourcesLoaded || dataSources.length === 0}
+              loading={!dataSourcesLoaded}
+              optionFilterProp="data-name"
+              showSearch
+              onChange={handleDataSourceChange}>
+              {map(dataSources, ds => (
+                <Select.Option key={`ds-${ds.id}`} value={ds.id} data-name={ds.name}>
+                  <img src={`/static/images/db-logos/${ds.type}.png`} width="20" alt={ds.name} />
+                  <span>{ds.name}</span>
+                </Select.Option>
+              ))}
+            </Select>
+          </div>
+          <div className="editor__left__schema">
```

Comment:
@gabrieldutra @arikfr Here's another place where it would be great to add an additional `DynamicComponent` so we can display some metadata about the query's data source (version and documentation link).

Right now we have to have a fork commit to display our React components like this: https://github.com/mozilla/redash/commit/4a82562e0539281064f9468ef4afd841084e58a0. Their actual code lives in our Redash extension library redash-stmo (e.g. [link](https://github.com/mozilla/redash-stmo/blob/master/src/redash_stmo/data_sources/link/bundle/datasource_link.js)).